pkg/util/proxyutil/reverse_proxy.go

```diff
@@ -0,0 +1,135 @@
+package proxyutil
+
+import (
+	"context"
+	"errors"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+
+	glog "github.com/grafana/grafana/pkg/infra/log"
+)
+
+// ReverseProxyOption reverse proxy option to configure a httputil.ReverseProxy.
+type ReverseProxyOption func(*httputil.ReverseProxy)
+
+// NewReverseProxy creates a new httputil.ReverseProxy with sane default configuration.
+func NewReverseProxy(logger glog.Logger, director func(*http.Request), opts ...ReverseProxyOption) *httputil.ReverseProxy {
+	if logger == nil {
+		panic("logger cannot be nil")
+	}
+
+	if director == nil {
+		panic("director cannot be nil")
+	}
+
+	p := &httputil.ReverseProxy{
+		FlushInterval: time.Millisecond * 200,
+		ErrorHandler:  errorHandler(logger),
+		ErrorLog:      log.New(&logWrapper{logger: logger}, "", 0),
+		Director:      director,
+	}
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	origDirector := p.Director
+	p.Director = wrapDirector(origDirector)
+
+	if p.ModifyResponse == nil {
+		// nolint:bodyclose
+		p.ModifyResponse = modifyResponse(logger)
+	} else {
+		modResponse := p.ModifyResponse
+		p.ModifyResponse = func(resp *http.Response) error {
+			if err := modResponse(resp); err != nil {
+				return err
+			}
+
+			// nolint:bodyclose
+			return modifyResponse(logger)(resp)
+		}
+	}
+
+	return p
+}
+
+// wrapDirector wraps a director and adds additional functionality.
+func wrapDirector(d func(*http.Request)) func(req *http.Request) {
+	return func(req *http.Request) {
+		d(req)
+		PrepareProxyRequest(req)
+
+		// Clear Origin and Referer to avoid CORS issues
+		req.Header.Del("Origin")
+		req.Header.Del("Referer")
+	}
+}
+
+// modifyResponse enforces certain constraints on http.Response.
+func modifyResponse(logger glog.Logger) func(resp *http.Response) error {
+	return func(resp *http.Response) error {
+		resp.Header.Del("Set-Cookie")
+		SetProxyResponseHeaders(resp.Header)
+		return nil
+	}
+}
+
+type timeoutError interface {
+	error
+	Timeout() bool
+}
+
+func errorHandler(logger glog.Logger) func(http.ResponseWriter, *http.Request, error) {
+	return func(w http.ResponseWriter, r *http.Request, err error) {
+		if errors.Is(err, context.Canceled) {
+			logger.Debug("Proxy request cancelled")
+			w.WriteHeader(http.StatusBadRequest)
```

Comment:
Is this where we change status for cancelled requests? Perhaps some comment about it :)
I think 499 _might_ be more appropriate as well. It's not standard but fairly common for proxies. not sure. 400 could be confused with invalid input which we care more about. 

https://httpstatus.in/499/

Reply:
> Is this where we change status for cancelled requests? 

Yes, that's correct. You don't think `logger.Debug("Proxy request cancelled")` is enough?

> I think 499 _might_ be more appropriate as well. It's not standard but fairly common for proxies. not sure. 400 could be confused with invalid input which we care more about.
> 
> https://httpstatus.in/499/

Yeah I did consider this status, but based on my minor read up sounded like this was only used by nginx. Have you seen this used in other places?

For the client/browser cancelling the request they'll never receive this status code - it's mainly for logs and metrics. Metrics for proxied request will need additional work outside of this, but one thought currently is that we shouldn't track request metrics for our proxies at all. If needed, we should track specific proxy metrics. Thoughts?

Reply:
I'll push an update with 499.

Also found another good source for 499 https://www.belugacdn.com/499-error-code/

Reply:
Done. Need to do some manual test to verify it works as expected